### ISSUE TITLE
Problem with multiple results std:tuple from Lua function

### ISSUE THREAD

**Issue by chenjiansysu** [SteveKChiu/lua-intf#71]
Hello, Steve.
Can you give me a hint?

CPP:

    LuaIntf::LuaRef require(context->state(), "require");
    require("main");
    LuaIntf::LuaRef main(context->state(), "main");
    std::string str;
    int errcode;
    std::tie(errcode, str) = main.call<std::tuple<int,std::string>>();

main.lua:

    function main()
        return 1, "test"
    end

MinGW 4.9.1 32bit under Win8 with Lua5.3

Unfortunately, I got a lua error: bad argument #1 (number expected, got function)...

**Comment by SteveKChiu** [comment]
It is a bug, but I am not sure how it get thru, I must be missing something.
Anyway, thanks for report the problem.

**Comment by chenjiansysu** [comment]
The problem still exists! But It works find when i modify the source code like this. 

    template <typename... R, typename... P>
    struct Call <std::tuple<R...>, P...>
    {
        static std::tuple<R...> invoke(lua_State* L, const LuaRef& f, P&&... args)
        {
            lua_pushcfunction(L, &LuaException::traceback);
            f.pushToStack();
            pushArg(L, std::forward<P>(args)...);
            //if (lua_pcall(L, sizeof...(P), sizeof...(R), -int(sizeof...(P) + sizeof...(R) + 2)) != LUA_OK) {
            if (lua_pcall(L, sizeof...(P), sizeof...(R), -int(sizeof...(P) + 2)) != LUA_OK) {
                lua_remove(L, -2);
                throw LuaException(L);
            }
            lua_remove(L, 1); // Remove the traceback function from the bottom of stack
            std::tuple<R...> ret;
            Result<sizeof...(R), R...>::fill(L, ret);
            lua_pop(L, int(sizeof...(R) + 1));
            return ret;
        }
    };

However, I am  not sure if this is correct. Thank you very much for your reply.

**Comment by chenjiansysu** [comment]
Another solution：

    template <size_t N, typename... P>
    struct Result
    {
        static void fill(lua_State* L, std::tuple<P...>& args)
        {
            //fillResult(L, sizeof...(P) - N + 1, std::get<sizeof...(P) - N>(args));
            fillResult(L, -N, std::get<sizeof...(P) - N>(args));
            Result<N - 1, P...>::fill(L, args);
        }
    };

**Comment by SteveKChiu** [comment]
Your first modification is exactly what I have done in the new commit, have you updated the code?

**Comment by chenjiansysu** [comment]
I had updated the code.  But the problem still exists without  the following code:

    lua_remove(L, 1); // Remove the traceback function from the bottom of stack
The function of Result<sizeof...(R), R...>::fill(L, ret) will get the traceback function which at the bottom of the lua stack as the first return value, so I got the lua error(number expected, got function).

After the second modification, Result<sizeof...(R), R...>::fill(L, ret) will get the return value from the top of lua stack.
